### PR TITLE
(BOLT-152) Default to secure ssh authentication

### PR DESCRIPTION
**Pull request by joshcooper**
This commit modifies bolt to use the `secure` verifier by default. The
connection will be allowed if the remote host has an entry in our known_hosts
file and the key presented by the host matches our local copy.

Add a command line option (either -k or --insecure) to fallback to the old
lenient behavior. The other net-ssh verifiers `null` and `strict` are not
supported.

### CODE REVIEW

File: lib/bolt/cli.rb [hunk 1]
```diff
@@ -130,6 +130,10 @@ def create_option_parser(results)
                 "Parameters to a task or plan") do |params|
           results[:task_options] = parse_params(params)
         end
+        opts.on('-k', '--insecure',
+                "Whether to connect insecurely ") do |insecure|
+          Bolt.config[:insecure] = insecure
+        end
```
Comment (joshcooper) [inline lib/bolt/cli.rb]: need to add a cli test

File: lib/bolt/result.rb [hunk 2]
```diff
@@ -25,6 +25,16 @@ def success?
     end
   end
 
+  class ErrorResult < Result
+    def to_h
+      { 'value' => @message }
+    end
```
Comment (joshcooper) [inline lib/bolt/result.rb]: This seems to be sufficient for the `plan` case:

```
$ bundle exec bolt plan run sample::single_task nodes=qkjbgx7gwjt2dqj.delivery.puppetlabs.net --modules spec/fixtures/modules/
ExecutionResult({'qkjbgx7gwjt2dqj.delivery.puppetlabs.net' => 'Host key verification failed for qkjbgx7gwjt2dqj.delivery.puppetlabs.net: fingerprint ac:8c:26:9b:20:d6:71:2d:cd:a9:1f:40:cb:a9:1f:98 is unknown for "qkjbgx7gwjt2dqj.delivery.puppetlabs.net,10.32.123.243"'})
```

But I don't know if that's sufficient for all cases?

Comment (pcarlisle) [reply lib/bolt/result.rb]: but we want it to be an error, this is making it a success in puppet

Comment (pcarlisle) [reply lib/bolt/result.rb]: should be something like
```ruby
{
  'error' => {
    'kind' => 'puppetlabs.tasks/connection-error',
    'issue_code' => 'HOST_KEY_SOMETHING',
    'msg' => "Host key verification failed",
    'details' => { ?? }
  }
}
```

Comment (joshcooper) [reply lib/bolt/result.rb]: The `success?` method returns false below, ah but right, the hash contents need to be error like. Does it need to contain `_error` and `_output`, or only `_error`?

Comment (joshcooper) [reply lib/bolt/result.rb]: ah, right you were talking about the issue code and parameterizing it.

File: lib/bolt/node/exception.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+module Bolt
+  class Node
+    class Exception < RuntimeError; end
```
Comment (pcarlisle) [inline lib/bolt/node/exception.rb]: the name should indicate what this is for somehow

File: lib/bolt/node/ssh.rb [hunk 4]
```diff
@@ -6,13 +6,24 @@
...
+        non_interactive: true
+      }
+
       options[:port] = @port if @port
       options[:password] = @password if @password
+      options[:verify_host_key] = if @insecure
+                                    Net::SSH::Verifiers::Lenient.new
+                                  else
+                                    Net::SSH::Verifiers::Secure.new
+                                  end
```
Comment (joshcooper) [inline lib/bolt/node/ssh.rb]: This overrides whatever is in `~/.ssh/config` which I don't think is desirable, e.g. if it contains `StrictHostKeyChecking no` for the host/domain you're trying to connect to.

Comment (joshcooper) [reply lib/bolt/node/ssh.rb]: `net-ssh` doesn't parse the `StrictHostKeyChecking` value for a host, e.g. given:

```
  StrictHostKeyChecking no
```

net-ssh doesn't load the value:

```
Net::SSH::Config.for(@host)
{:auth_methods=>["none", "publickey", "password", "keyboard-interactive"], :user=>"root", :keys=>["~/.ssh/id_rsa", "~/.ssh/josh_puppetlabs"], :send_env=>[/^LANG$/, /^LC_.*$/]}
```

**Comment by joshcooper** [comment]
The plan result now indicates an error:

```
$ bx bolt plan run sample::single_task nodes=jgpua7467pa7jqz.delivery.puppetlabs.net --modules spec/fixtures/modules
ExecutionResult({'jgpua7467pa7jqz.delivery.puppetlabs.net' => Error({'message' => 'Host key verification failed for jgpua7467pa7jqz.delivery.puppetlabs.net: fingerprint ac:8c:26:9b:20:d6:71:2d:cd:a9:1f:40:cb:a9:1f:98 is unknown for "jgpua7467pa7jqz.delivery.puppetlabs.net,10.32.124.18"', 'kind' => 'puppetlabs.tasks/connect-error', 'issue_code' => 'NODE_ERROR'})})
```

**Comment by pcarlisle** [comment]
I think the error code is supposed to be more specific than the kind. I was under the impression that it should be specific to the exact error, equivalent to the message but not subject to interpolation or translation.

@hlindberg

**Comment by joshcooper** [comment]
I switched `issue_code` to match the exception class `CONNECT_ERROR`, while the `kind` remains more generic as `puppetlabs.tasks/node-error`.

File: lib/bolt/node/errors.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      end
+
+      def issue_code
+        'NODE_ERROR'
+      end
+    end
+
+    class ConnectError < BaseError
+      def issue_code
+        'CONNECT_ERROR'
```
Comment (joshcooper) [inline lib/bolt/node/errors.rb]: make this more specific, e.g. AUTH_ERROR, HOST_KEY_ERROR, etc

**Pull request merged**